File: QytetetRuby/pkg/QytetetRuby-0.0.1/lib/jugador.rb
#encoding: utf-8
module ModeloQytetet
	
	class Jugador
		attr_accessor :cartaLibertad, :casillaActual, :encarcelado
		attr_reader :nombre, :propiedades, :saldo
		
		def initialize(nombre, saldo, encarcelado, casillaActual, cartaLibertad, propiedades)
			@nombre=nombre
			@saldo=saldo
			@encarcelado=encarcelado
			@casillaActual=casillaActual
			@cartaLibertad=cartaLibertad
			@propiedades=propiedades
		end
		
		def self.nuevo(nombre)
			new(nombre, 7500, false, Casilla.new(0, 1000, TipoCasilla::PARKING), nil, Array.new)
		end
		
		def self.copia(otroJugador)
			new(otroJugador.nombre, otroJugador.saldo, otroJugador.encarcelado, otroJugador.casillaActual, 
					 otroJugador.cartaLibertad, otroJugador.propiedades)
		end
		
		def cancelarHipoteca(titulo)
			costeHipotecar = titulo.calcularCosteCancelar
			tengoSaldo = tengoSaldo(costeHipotecar)
			
			if (tengoSaldo)
				modificarSaldo(-costeHipotecar)
				titulo.cancelarHipoteca
			end
			
			return tengoSaldo
		end
		
		def comprarTituloPropiedad
			comprado = false
			costeCompra = @casillaActual.coste
			
			if (costeCompra < @saldo)
				titulo = @casillaActual.asignarPropietario(self)
				comprado = true
				
				@propiedades << titulo
				modificarSaldo(-costeCompra)
			end
			
			return comprado
		end
		
		def convertirme(fianza)
			Especulador.copia(self, fianza)
		end

		def cuantasCasasHotelesTengo
			total = 0
			
			for propiedad in @propiedades
				total += propiedad.numCasas + propiedad.numHoteles
			end
			
			return total
		end
		
		def deboIrACarcel
			return !tengoCartaLibertad
		end
		
		def deboPagarAlquiler
			deboPagar = false
			titulo = @casillaActual.titulo
			esDeMiPropiedad = esDeMiPropiedad(titulo)
			
			if (!esDeMiPropiedad)
				tienePropietario = titulo.tengoPropietario
				
				if (tienePropietario)
					encarcelado = titulo.propietarioEncarcelado
					
					if (!encarcelado)
						estaHipotecada = titulo.hipotecada
						
						if (!estaHipotecada)
							deboPagar = true
						end
					end
				end
			end
			
			return deboPagar
		end
			
		def devolverCartaLibertad
			@cartaLibertad=nil
			
			Sorpresa.new("Te sobra dinero para pagar la fianza, no necesitas estar en la cárcel como los putos pobres :)", 0, TipoSorpresa::SALIRCARCEL)
		end
		
		def edificarCasa
			
		end
		
		def edificarHotel
			
		end
		
		def edificarCasa(titulo)
			edificada = false
			
			if (puedoEdificarCasa(titulo))
				costeEdificarCasa = titulo.precioEdificar
				tengoSaldo = tengoSaldo(costeEdificarCasa)
				
				if (tengoSaldo)
					titulo.edificarCasa()
					modificarSaldo(-costeEdificarCasa)
					edificada = true
				end
			end
			
			return edificada
		end
		
		def edificarHotel(titulo)
			edificada = false
			
			if (puedoEdificarHotel(titulo))
				costeEdificarHotel = titulo.precioEdificar
				tengoSaldo = tengoSaldo(costeEdificarHotel)
				
				if (tengoSaldo)
					titulo.edificarHotel()
					modificarSaldo(-costeEdificarHotel)
					edificada = true
				end
			end
			
			return edificada
		end
		
		def eliminarDeMisPropiedades(titulo)
			@propiedades.delete(titulo)
			@propietario = nil
		end
		
		def esDeMiPropiedad(titulo)
			soyPropietario = false
			
			for propiedad in @propiedades
				if(propiedad == titulo)
					soyPropietario = true;
				end
			end
		end
		
		#def estoyEnCalleLibre
		
		#def getCartaLibertad
		
		def hipotecarPropiedad(titulo)
			costeHipoteca = titulo.hipotecar
			modificarSaldo(costeHipoteca)
		end
		
		def irACarcel(casilla)
			@casillaActual = casilla
			@encarcelado = true
		end
		
		def modificarSaldo(cantidad)
			@saldo += cantidad
		end
		
		def obtenerCapital
			capital = @saldo
			
			for propiedad in @propiedades
				capital += propiedad.precioCompra + (propiedad.numCasas + propiedad.numHoteles) * propiedad.precioEdificar
				if(propiedad.hipotecada)
					capital -= propiedad.hipotecaBase
				end
			end
			
			return capital
		end
		
		def obtenerPropiedades(hipotecada)
			hipotecadas = Array.new
			
			for propiedad in @propiedades
				if (propiedad.hipotecada == hipotecada)
					hipotecadas << propiedad
				end
			end
		end
		
		def pagarAlquiler
			costeAlquiler = @casillaActual.pagarAlquiler()
			
			modificarSaldo(-costeAlquiler)
		end
		
		def pagarImpuesto
			@saldo -= @casillaActual.coste
		end
		
		def pagarLibertad(cantidad)
			tengoSaldo = tengoSaldo(cantidad)
			
			if (tengoSaldo)
				@encarcelado = false
				modificarSaldo(-cantidad)
			end
		end
		
		def puedoEdificarCasa(titulo)
			titulo.numCasas < 4
		end
		
		def puedoEdificarHotel(titulo)
			titulo.numCasas == 4 && titulo.numHoteles < 4
		end
		
		def tengoCartaLibertad
			cartaLibertad != nil
		end
		
		def tengoSaldo(cantidad)
			@saldo > cantidad
		end
		
		def venderPropiedad(casilla)
			titulo = casilla.titulo
			eleminarDeMisPropiedades(titulo)
			
			precioVenta = calcularPrecioVenta
			modificarSaldo(precioVenta)
		end
		
		#Override
		def <=> (otroJugador)
			otroJugador.ObtenerCapital <=> obtenerCapital
		end


		#Override
		def to_s()
			"Jugador{nombre=#{@nombre}, saldo=#{@saldo}, capital=#{obtenerCapital}, encarcelado=#{@encarcelado}, casillaActual=#{@casillaActual}, cartaLibertad=#{@cartaLibertad}, propiedades=#{@propiedades}}"
		end
		
		def to_s_sinpropiedad()
			"Jugador{nombre=#{@nombre}, saldo=#{@saldo}, capital=#{obtenerCapital}, encarcelado=#{@encarcelado}, casillaActual=#{@casillaActual}, cartaLibertad=#{@cartaLibertad}}"
		end
		
		#private :eliminarDeMisPropiedad, :esDeMiPropiedad
		#protected :convertirme, :deboIrACarcel, :copia, :pagarImpuesto, :puedoEdificarCasa, :puedoEdificarHotel, :tengoSaldo
	end
end

File: QytetetJava/src/modeloqytetet/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

import java.util.ArrayList;

/**
 *
 * @author jakerxd
 */
public class Jugador implements Comparable{
	
	private String nombre;
	private int saldo;
	private boolean encarcelado;
	private Casilla casillaActual;
	private Sorpresa cartaLibertad;
	private ArrayList<TituloPropiedad> propiedades;
	
	
	Jugador(String nombre)
	{
		this.nombre = nombre;
		saldo = 7500;
		encarcelado = false;
		casillaActual = new OtraCasilla(0, 500, TipoCasilla.PARKING);
		cartaLibertad = null;
		propiedades = new ArrayList();
	}
	
	Jugador(Jugador otroJugador)
	{
		nombre = otroJugador.nombre;
		saldo  = otroJugador.saldo;
		encarcelado = otroJugador.encarcelado;
		casillaActual = otroJugador.casillaActual;
		cartaLibertad = otroJugador.cartaLibertad;
		propiedades = otroJugador.propiedades;
	}

	boolean getEncarcelado() {
		return encarcelado;
	}

	String getNombre() {
		return nombre;
	}

	public int getSaldo() {
		return saldo;
	}
	
	boolean cancelarHipoteca(TituloPropiedad titulo)
	{
		int costeHipotecar = titulo.calcularCosteCancelar();
		boolean tengoSaldo = tengoSaldo(costeHipotecar);
				
		if (tengoSaldo)
		{
			modificarSaldo(-costeHipotecar);
			titulo.cancelarHipoteca();
		}
			
		return tengoSaldo;
	}
	
	boolean comprarTituloPropiedad()
	{
		boolean comprado = false;
		int costeCompra = casillaActual.getCoste();
		
		if (costeCompra < saldo)
		{
			TituloPropiedad titulo = casillaActual.asignarPropietario(this);
			
			comprado = true;
			propiedades.add(titulo);
			modificarSaldo(-costeCompra);
		}
		
		return comprado;
	}
	
	protected Especulador convertirme(int fianza)
	{
		return new Especulador(this, fianza);
	}
	
	int cuantasCasasHotelesTengo()
	{
		int total = 0;
		
		for (int i=0; i<propiedades.size(); i++)
			total += propiedades.get(i).getNumCasas() + propiedades.get(i).getNumHoteles();
		
		return total;
	}
	
	protected boolean deboIrACarcel()
	{
		return !tengoCartaLibertad();
	}
	
	boolean deboPagarAlquiler()
	{
		TituloPropiedad titulo = casillaActual.getTitulo();
		boolean esDeMiPropiedad = esDeMiPropiedad(titulo);
		boolean deboPagar = false;
		
		if (!esDeMiPropiedad)
		{
			boolean tienePropietario = titulo.tengoPropietario();
			if (tienePropietario)
			{
				boolean encarcelado = titulo.propietarioEncarcelado();
				if (!encarcelado)
				{
					boolean estaHipotecada = titulo.getHipotecada();
					if (!estaHipotecada)
						deboPagar = true;
				}
			}
		}
		
		return deboPagar;
	}
	
	Sorpresa devolverCartaLibertad()
	{
		cartaLibertad = null;
		
		return new Sorpresa("Te sobra dinero para pagar la fianza, no necesitas estar en la cárcel como los putos pobres :)", 0, TipoSorpresa.SALIRCARCEL);
	}
	
	boolean edificarCasa(TituloPropiedad titulo)
	{
		boolean edificada = false;
		
		if (puedoEdificarCasa(titulo))
		{
			int costeEdificarCasa = titulo.getPrecioEdificar();
			boolean tengoSaldo = tengoSaldo(costeEdificarCasa);
			
			if (tengoSaldo)
			{
				titulo.edificarCasa();
				modificarSaldo(-costeEdificarCasa);
				edificada = true;
			}
		}
		
		return edificada;
	}
	
	boolean edificarHotel(TituloPropiedad titulo)
	{
		boolean edificada = false;
		
		if (puedoEdificarHotel(titulo))
		{
			int costeEdificarHotel = titulo.getPrecioEdificar();
			boolean tengoSaldo = tengoSaldo(costeEdificarHotel);
			
			if (tengoSaldo)
			{
				titulo.edificarHotel();
				modificarSaldo(-costeEdificarHotel);
				edificada = true;
			}
		}
		
		return edificada;
	}
	
	private void eliminarDeMisPropiedades(TituloPropiedad titulo)
	{
		propiedades.remove(titulo);
		titulo.setPropietario(null);
	}
	
	private boolean esDeMiPropiedad(TituloPropiedad titulo)
	{
		boolean soyPropietario = false;
		
		for(int i=0; i<propiedades.size(); i++)
			if(titulo == propiedades.get(i))
				soyPropietario = true;
		
		return soyPropietario;
	}
	
	//boolean estoyEnCalleLibre(){}
	
	Sorpresa getCartaLibertad(){
		return cartaLibertad;
	}
	
	public Casilla getCasillaActual(){
		return casillaActual;
	}
	
	ArrayList<TituloPropiedad> getPropiedades(){
		return propiedades;
	}
	
	void hipotecarPropiedad(TituloPropiedad titulo)
	{
		int costeHipoteca = titulo.hipotecar();
		
		modificarSaldo(costeHipoteca);
	}
	
	void irACarcel(Casilla casilla)
	{
		setCasillaActual(casilla);
		setEncarcelado(true);
	}
	
	void modificarSaldo(int cantidad)
	{
		saldo += cantidad;
	}
	
	int obtenerCapital()
	{
		int capital = saldo;
		TituloPropiedad titulo;
		
		for (int i=0; i<propiedades.size(); i++)
		{
			titulo = propiedades.get(i);
			capital += titulo.getPrecioCompra();
			capital += titulo.getPrecioEdificar() * (titulo.getNumCasas() + titulo.getNumHoteles());
			if(titulo.getHipotecada())
				capital -= titulo.getHipotecaBase();
		}
		
		return capital;
	}
	
	ArrayList<TituloPropiedad> obtenerPropiedades(boolean estadoHipotecada)
	{
		ArrayList<TituloPropiedad> estado = new ArrayList();
		
		for(int i=0; i<propiedades.size(); i++)
		{
			if (propiedades.get(i).getHipotecada() == estadoHipotecada)
				estado.add(propiedades.get(i));
		}
		
		return estado;
	}
	
	void pagarAlquiler()
	{
		int costeAlquiler = ( (Calle) casillaActual).pagarAlquiler();
		
		modificarSaldo(-costeAlquiler);
	}
	
	protected void pagarImpuesto()
	{
		saldo -= casillaActual.getCoste();
	}
	
	void pagarLibertad(int cantidad)
	{
		boolean tengoSaldo = tengoSaldo(cantidad);
		
		if (tengoSaldo)
		{
			setEncarcelado(false);
			modificarSaldo(-cantidad);
		}
	}
	
	protected boolean puedoEdificarCasa(TituloPropiedad titulo)
	{
		return (titulo.getNumCasas() < 4);
	}
	
	protected boolean puedoEdificarHotel(TituloPropiedad titulo)
	{
		return (titulo.getNumCasas() == 4 && titulo.getNumHoteles() < 4);
	}
	
	void setCartaLibertad(Sorpresa carta){
		cartaLibertad = carta;
	}
	
	void setCasillaActual(Casilla casilla){
		casillaActual = casilla;
	}
	
	void setEncarcelado(boolean encarcelado){
		this.encarcelado = encarcelado;
	}
	
	void setEncarcelado(){
		this.encarcelado = false;
	}
	
	boolean tengoCartaLibertad(){return cartaLibertad != null;}
	
	protected boolean tengoSaldo(int cantidad)
	{
		return (saldo > cantidad);
	}
	
	void venderPropiedad(Casilla casilla)
	{
		TituloPropiedad titulo = casilla.getTitulo();
		int precioVenta = titulo.calcularPrecioVenta();
		
		eliminarDeMisPropiedades(titulo);
		modificarSaldo(precioVenta);
	}
	
	@Override
	public int compareTo (Object otroJugador)
	{
		return obtenerCapital() - ((Jugador)otroJugador).obtenerCapital();
	}
	
	@Override
	public String toString() 
	{
		if(cartaLibertad == null)
			return "Jugador{nombre=" + nombre + ", saldo=" + saldo + ", capital=" + obtenerCapital() + ", encarcelado=" + encarcelado + ", casillaActual=" + casillaActual.toString() + ", propiedades=" 
					+ propiedades.toString();
		else
			return "Jugador{nombre=" + nombre + ", saldo=" + saldo + ", capital=" + obtenerCapital() + ", encarcelado=" + encarcelado + ", casillaActual=" + casillaActual.toString() + ", cartaLibertad=" 
					+ cartaLibertad.toString() + ", propiedades=" + propiedades.toString();
	}
}